File: apps/api/app/schemas/project.py
from datetime import date, datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.base import to_camel


class ProjectStatus(StrEnum):
    PLANNING = "planning"
    ACTIVE = "active"
    PAUSED = "paused"
    COMPLETED = "completed"
    ARCHIVED = "archived"


class ProjectPriority(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class ProjectBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    status: ProjectStatus = ProjectStatus.PLANNING
    priority: ProjectPriority = ProjectPriority.MEDIUM
    start_date: date | None = None
    due_date: date | None = None

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class ProjectCreate(ProjectBase):
    workspace_id: UUID
    client_id: UUID


class ProjectUpdate(BaseModel):
    workspace_id: UUID | None = None
    client_id: UUID | None = None
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    status: ProjectStatus | None = None
    priority: ProjectPriority | None = None
    start_date: date | None = None
    due_date: date | None = None

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class ProjectRead(ProjectBase):
    id: UUID
    workspace_id: UUID
    client_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        alias_generator=to_camel,
        from_attributes=True,
        populate_by_name=True,
    )

File: apps/web/types/project.ts
export type ProjectStatus =
  | "planning"
  | "active"
  | "paused"
  | "completed"
  | "archived";

export type ProjectPriority = "low" | "medium" | "high";

export type Project = {
  id: string;
  workspaceId: string;
  clientId: string;
  name: string;
  description: string | null;
  status: ProjectStatus;
  priority: ProjectPriority;
  startDate: string | null;
  dueDate: string | null;
  createdAt: string;
  updatedAt: string;
};

export type ProjectPayload = {
  workspaceId: string;
  clientId: string;
  name: string;
  description?: string | null;
  status: ProjectStatus;
  priority: ProjectPriority;
  startDate?: string | null;
  dueDate?: string | null;
};

export type ProjectFilters = {
  workspaceId?: string;
  clientId?: string;
  status?: ProjectStatus;
};
